### PR TITLE
Populate first name, last name, and email, if available.

### PR DESCRIPTION
**Pull request by adelton**
This pull request builds on top of https://github.com/theforeman/foreman/pull/967 to populate user attributes of newly created users.

### CODE REVIEW

**Comment by adelton** [comment]
Rebased on top of develop (and now merged https://github.com/theforeman/foreman/pull/967 ): 41d60e82844eac56ad6e24e14c3476d953776b02.

File: app/services/sso/apache.rb [hunk 1]
```diff
@@ -20,9 +20,23 @@ def support_fallback?
...
+    }
     def authenticated?
       return false unless (self.user = request.env[CAS_USERNAME])
-      return false unless User.find_or_create_external_user(self.user, Setting['authorize_login_delegation_auth_source_user_autocreate'])
+      attrs = { :login => self.user }
+      ENV_TO_ATTR_MAPPING.each do |k, v|
+        if request.env.has_key?(k)
+          attrs[v] = request.env[k].dup.force_encoding(Encoding::UTF_8)
+          if not attrs[v].valid_encoding?
+            attrs[v] = attrs[v].encode(Encoding::ISO_8859_1, Encoding::UTF_8, {:invalid => :replace, :replace => '-'}).force_encoding(Encoding::UTF_8)
```
Comment (domcleal) [inline app/services/sso/apache.rb]: I don't think this is valid on Ruby 1.8, you might need to look into iconv instead.  I seem to remember Encoding is 1.9 only.

**Comment by domcleal** [comment]
Could you look at the tests as well?  I'll run them now: [test] but I expect the ones in test/unit/user_test.rb for User.find_or_create_external_user are going to fail as the method signature's changed to accept the hash.  I'd say we also need updated tests in two places:

1) test/unit/user_test.rb adding a new find_or_create_external_user test to ensure it also sets first name, surname etc on the newly created user when given a hash of attributes.

2) test/unit/sso/apache_test.rb adding a new test of .authenticated? which sets the appropriate request.env["REMOTE_USER_FIRSTNAME"] variable (ensuring it's ISO 8859-1 too) and can then check it calls the user method using:

```
User.expects(:find_or_create_external_user).with({...}).returns(true)
```

or similar.

**Comment by domcleal** [comment]
This currently only populates the info on the initial creation (which I think is the same for LDAP auth sources too), is that intended or do you want to update every time?

File: app/services/sso/apache.rb (hunk 1 above)
Comment (adelton) [inline app/services/sso/apache.rb]: So should there be an if to do iconv on 1.8 and Encoding on 1.9? On what OS/version does Foreman use 1.8? Can we make the Encoding conditional on 1.9, not supporting (== leaving whatever is there, hopefully US-ASCII) UTF-8 on installations on ruby 1.8?

**Comment by adelton** [comment]
As for the test, I think I'll ask Marek H. for them -- he did a good job for https://github.com/theforeman/foreman/pull/967.

**Comment by adelton** [comment]
As for updates not done -- this is intentional, the code mimics the plain internal LDAP behaviour as closely as possible. So at this point it only populates the attributes only upon the initial creation of the user record. When/if the LDAP behaviour changes, this will need to change as well.

File: app/services/sso/apache.rb (hunk 1 above)
Comment (domcleal) [inline app/services/sso/apache.rb]: 1.8's currently used on Debian and Ubuntu.  I recall seeing user problems with US-ASCII data getting stored from LDAP(?), so this is definitely welcome.

I think a conditional's the best way of doing it, e.g.

<pre>
if attrs[v].respond_to? :force_encoding
  # existing code
else
  attrs[v] = Iconv.new('UTF-8//IGNORE', 'UTF-8').iconv(attrs[v]) rescue attrs[v]
end
</pre>


That's roughly equivalent to the force_encoding bit I think, skipping the actual character encoding change you're doing after (useful, but that's a nice to have for 1.8 at this stage I'd say, unless you want to check how to do it with iconv).

Comment (domcleal) [inline app/services/sso/apache.rb]: Oh, and at the top of this source file you'd need to add:

```
require 'iconv' if RUBY_VERSION.start_with?('1.8.')
```

Comment (lzap) [inline app/services/sso/apache.rb]: From supported OSes currently only Debians I think. This should not be hard, just create some helper method that will use `Encoding` or `Iconv` class depending on the Ruby version. It should have been definitely covered it with a unit test.

Comment (adelton) [inline app/services/sso/apache.rb]: The 1.8 support addressed in 9a35f6d2670b92563c9cc1647e7f413cfe4ecab6.

**Comment by adelton** [comment]
I have now updated the description of http://projects.theforeman.org/issues/3528 with steps to see the feature working with IPA server.

**Comment by ohadlevy** [comment]
@ares mind having a look? tests are failing.

**Comment by ares** [comment]
I sent a fix for Ruby 1.9.3 in https://github.com/adelton/foreman/pull/5

**Comment by adelton** [comment]
> I sent a fix for Ruby 1.9.3 in adelton#5

Thank you. Rebased as 210b9cdbd4e8818772567b8b0cbedd4f0ad95f83 on top of develop.

**Comment by domcleal** [comment]
Thanks for the contribution @adelton and for the tests @ares.  Merged as 8ffa0b9a for Foreman 1.4.0.

**Pull request closed without merging**